### ISSUE TITLE
torch.quantile on MPS doesn't sort values when dim is not None

### ISSUE THREAD

**Issue by oblander** [pytorch/pytorch#101878]
### 🐛 Describe the bug

MPS implementation of `torch.quantile` doesn't sort the input when there's the `dim` argument is not `None`, giving nonsense results.

```
import torch
device = torch.device("mps")
A = torch.randn(1000, 10)

#need to use "nearest", "lower", or "higher" interpolation methods
#"linear" and "midpoint" fail due to operator `aten::lerp.Tensor_out` not being implemented for MPS
#when dim=None, CPU and MPS give same result
torch.quantile(A, .5, interpolation = "nearest")
torch.quantile(A.to(device), .5, interpolation = "nearest")

#when dim is not None, MPS gives nonsense result
torch.quantile(A, .5, 0, interpolation = "nearest")
torch.quantile(A.to(device), .5, 0, interpolation = "nearest")

#MPS output is same as middle row of tensor
#so torch.quantile must just be pulling 500th index without sorting first
A[500]

#inelegant workaround is to sort the tensor before calling quantile
torch.quantile(torch.sort(A.to(device), dim = 0).values, .5, 0, interpolation = "nearest")
```

### Versions

PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.6
CMake version: version 3.22.1
Libc version: N/A

Python version: 3.11.3 (main, Apr 19 2023, 18:51:09) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.1
[pip3] torchmetrics==0.11.4
[conda] nomkl                     3.0                           0  
[conda] numpy                     1.24.3          py311hd0e4f0d_0  
[conda] numpy-base                1.24.3          py311h947b413_0  
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torchmetrics              0.11.4                   pypi_0    pypi

**Comment by oblander** [comment]
Addendum: in addition to the values returned being incorrect, the gradients returned by the backward method for `torch.quantile` are weird on MPS (usually all zeros, but sometimes not). Example below.

```
import torch
device = torch.device("mps")
A = torch.randn(1000, 10, requires_grad = True)

#CPU: correctly returns non-zero gradients for 1 element in each column of A
out = torch.sum(torch.quantile(A, .5, 0, interpolation = "nearest"))
out.backward()
torch.where(A.grad != 0)

A.grad = None

#MPS: usually returns no non-zero elements
#some seeds give non-zero elements but in the wrong places (always in the first row)
out = torch.sum(torch.quantile(A.to(device), .5, 0, interpolation = "nearest"))
out.backward()
torch.where(A.grad != 0)
```

**Comment by qqaatw** [comment]
Hi, from the environment information you provide, It looks that you're running Python using Rosetta 2, i.e. showing `macOS-10.16-x86_64-i386-64bit`. Can you please try running PyTorch on Python built natively on arm platform and see if the incorrect results persist?

Results from MPS backend on x86_64 platform are known to be erroneous in some cases.

**Comment by oblander** [comment]
Hi again, I set up a new environment using an arm64 build of Python and both bugs persisted (doesn't sort before returning values; gradients are all zero).

Here's my updated environment information:

```
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.3 (main, Apr 19 2023, 18:49:55) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.1
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
```

**Comment by qqaatw** [comment]
Hi, thanks for confirming. I'm not able to reproduce as my macOS stays on 12.x while the native mps support of sort is from macOS 13 onward IIRC. Will let others take a look at this.

**Comment by oblander** [comment]
Thanks! In case it's helpful, `torch.sort` does appear to work as expected on MPS (even on the x86_64 build), so my inelegant workaround to calculate quantiles is:
```
A_sorted = torch.sort(A, 0).values
A_sorted[round(0.5*A.shape[0])]
```
where `0.5` can be replaced by arbitrary quantiles and `round` can be replaced by `floor`/`ceil` to replicate `lower`/`upper` interpolation methods instead of `nearest`, etc.

**Comment by oblander** [comment]
Hey @Az0s, glad you were able to reproduce on a different setup so I know it's not just my machine+build!

Hopefully someone on the PyTorch team can point towards how to help with a fix––I'm just an end user, not a dev, so I don't know what the procedure is. But it looks like #77764 might be a good reference to start?

As mentioned above, it looks like `torch.quantile` returns the `round(q*n)` index in the dimension being reduced (in the MRE, taking the middle row of each column), which would be correct if the input was sorted along this dimension. Given that it works fine for `dim = None`, my guess is that somehow the sorting only happens when the input is flattened. As far as I can tell, `torch.sort` *does* already work correctly on MPS, even when `dim` is not `None`, so hopefully it's a relatively easy fix?

**Comment by qqaatw** [comment]
@Az0s you're welcomed to contribute! Let me know if you need any help.

**Comment by Az0s** [comment]
Hi everyone,

Firstly, I want to express my gratitude to all of you who have been providing insightful information regarding this issue. 

I've made some progress in identifying the problem. When looking at `sorting.cpp`, I've discovered that a specific operation `self.unsqueeze(-1).transpose(wrapped_dim, -1).sort()` which runs when `dim` is provided and `dim != inputTensor.dim() - 1` seems to cause the problem when run on mps devices. Here is the code to reproduce the issue:

```❯ python                        
imPython 3.11.3 (main, Apr 19 2023, 18:49:55) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> device = torch.device("mps")
>>> C=torch.randn(10, 5)
>>> C.unsqueeze(-1).transpose(0, -1).sort()[0][0][0], C.to(device).unsqueeze(-1).transpose(0, -1).sort()[0][0][0]
(tensor([-1.8086, -1.5690, -1.1932, -0.8253, -0.1920, -0.1752,  0.1716,  0.5822, 0.6503,  0.8782]), tensor([-0.8253,  0.5822,  0.6503, -1.1932, -0.1752,  0.1716, -1.5690,  0.8782,-0.1920, -1.8086], device='mps:0'))
```

This error persists and can be reproduced on the torch python interface.

I'm planning to approach this by trying to resolve the torch.quantile() issue without delving into other mps operations, if possible.

Additionally, it seems that this issue may be related to an error resulting from `tensor.transpose()` when run on mps devices. I'll look into whether there are related issues already reported. If I don't find any, I may consider opening a new one.

Lastly, I was wondering if any of you know a more efficient way to debug than modifying a .cpp file in ATen and rebuilding the entire Pytorch project to test if the changes work. Is there a static debugging method, for instance?

**Comment by qqaatw** [comment]
Hi, I guess (as I con't reproduce) the root cause is because `t.transpose()` and `t.unsqueeze()` produce a strided view that the `sort()` mps implementation doesn't handle well. You can add a contiguous clone after `transpose` to verify whether the output becomes correct, like so: 

```python
C.unsqueeze(-1).transpose(0, -1).clone(memory_format=torch.contiguous_format).sort()
```

If the above is the case, then it's tricky to resolve this issue exhaustively as not only this op is having this issue, and it requires some re-design at high level to address properly.

For debugging, I'm not quite sure about the static debugging method you were referring to, but the tip is to make use of `ccache`, `ninja`, and PCH, to cache and speed-up the compilation. You can take a look at `CONTRIBUTING.md` for more information.

**Comment by Az0s** [comment]
> Hi, I guess (as I can't reproduce) the root cause is `t.transpose()` and `t.unsqueeze()` produce a strided view that the `sort()` mps implementation doesn't handle well. You can add a contiguous clone after `transpose` to verify whether the output becomes correct, like so:
> 
> ```python
> C.unsqueeze(-1).transpose(0, -1).clone(memory_format=torch.contiguous_format).sort()
> ```
> 
> If the above is the case, then it's tricky to resolve this issue exhaustively as not only this op is having this issue, and it requires some re-design at high level to address properly.
> 
> For debugging, I'm not quite sure about the static debugging method you were referring to, but the tip is to make use of `ccache`, `ninja`, and PCH, to cache and speed-up the compilation. You can take a look at `CONTRIBUTING.md` for more information.

Thank you for the reply. I can confirm that the root cause is `t.transpose()` and `t.unsqueeze()` producing a strided view that the `sort()` MPS implementation doesn't handle well.

To fix this issue, I made some modifications to ensure that the input tensor is contiguous before passing it to the MPS. I have submitted a pull request with the specific solution. Could you please take a look and let me know if this is a proper solution? I would appreciate any advice you have.

#102524 

```diff
--- a/aten/src/ATen/native/mps/operations/Sort.mm
+++ b/aten/src/ATen/native/mps/operations/Sort.mm
@@ -29,10 +29,15 @@ TORCH_IMPL_FUNC(sort_stable_out_mps)
   bool macOS13_3_plus = is_macos_13_or_newer(MacOSVersion::MACOS_VER_13_3_PLUS);
   MPS_CHECK_INT64_OP_SUPPORTED(self, macOS13_3_plus, "sort_stable_out");
 
-  values.copy_(self);
-  // check if self is scalar
-  dim = maybe_wrap_dim(dim, self.dim(), true);
-  if (self.dim() == 0 && self.numel() == 1) {
+
+  // issue #101878: MPS might give a wrong sorted tensor when dealing with a strided view tensor.
+  // create a proxy tensor to temporarily address the issue
+  Tensor input_contiguous = self.contiguous();
+  Tensor values_contiguous = self.contiguous();
+
+  // check if input_contiguous is scalar
+  dim = maybe_wrap_dim(dim, input_contiguous.dim(), true);
+  if (input_contiguous.dim() == 0 && input_contiguous.numel() == 1) {
     indices.zero_();
     return;
   }
```

I should note that while this fix resolves the issue with `torch.quantile()`, the result of the gradients is still abnormal. I will investigate this further and make additional fixes if necessary.
